### PR TITLE
Log in Datadog When Sidekiq Worker Fails for Final Time

### PR DESCRIPTION
**Pull request by mstruve**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
Any time a worker has failed and will no longer retry we want to be notified bc manual intervention will be needed to either get the job to succeed. Once this is merged I will setup a monitor in Datadog to track when this number goes above 1. 

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/5#card-32687354

## Added to documentation?
- [x] no documentation needed

alt_text

### CODE REVIEW

**Comment by mstruve** [comment]
LOL yeah I noticed that too! 

I want to get more people familiar with this infrastructure so trying to mix up who I tag on PRs. Feel free to ask questions about anything you want! Hoping to help spread the knowledge!

File: config/initializers/sidekiq.rb [hunk 1]
```diff
@@ -24,4 +24,8 @@ def stringify_keys(other)
   config.server_middleware do |chain|
     chain.add Sidekiq::HoneycombMiddleware
   end
+
+  config.death_handlers << lambda do |job, _ex|
```
Comment (vaidehijoshi) [inline config/initializers/sidekiq.rb]: what is `_ex` here? do we _have_ to pass it in? if so, we might want to call out what it is.

File: spec/labor/sidekiq/worker_retries_exhausted_reporter_spec.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+require "rails_helper"
+
+describe Sidekiq::WorkerRetriesExhaustedReporter, type: :labor do
+  it "increments worker retries exhausted in Datadog" do
+    allow(DataDogStatsClient).to receive(:increment)
+    described_class.report_final_failure("class" => "TempWorker")
```
Comment (vaidehijoshi) [inline spec/labor/sidekiq/worker_retries_exhausted_reporter_spec.rb]:
```suggestion
    described_class.report_final_failure(class: TempWorker)
```
do you know if this new syntax would actually work? i feel like the hashrocket syntax is outdated, so i wonder if we could use this one instead?

File: app/labor/sidekiq/worker_retries_exhausted_reporter.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module Sidekiq
```
Comment (rhymes) [inline app/labor/sidekiq/worker_retries_exhausted_reporter.rb]: as this is part of the initializer, shouldn't it be in `lib/sidekiq` with the other custom middleware?

File: app/labor/sidekiq/worker_retries_exhausted_reporter.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  class WorkerRetriesExhaustedReporter
+    def self.report_final_failure(job)
```
Comment (rhymes) [inline app/labor/sidekiq/worker_retries_exhausted_reporter.rb]: I'm quite sure without the explanation you added in the PR someone (aka me) is going to ask what is this for in a few weeks :D I think it's worth adapting the PR description as a comment to this method

File: config/initializers/sidekiq.rb (hunk 1 above)
Comment (rhymes) [reply config/initializers/sidekiq.rb]: I think adding a comment linking to this https://github.com/mperham/sidekiq/wiki/Error-Handling#death-notification would be helpful

File: spec/labor/sidekiq/worker_retries_exhausted_reporter_spec.rb (hunk 2 above)
Comment (mstruve) [reply spec/labor/sidekiq/worker_retries_exhausted_reporter_spec.rb]: The reason I made this a string is bc that is how the actual arguments will be passed to the middleware. All job arguments are serialized when using Sidekiq so they will be strings when we parse them.

File: app/labor/sidekiq/worker_retries_exhausted_reporter.rb (hunk 3 above)
Comment (mstruve) [reply app/labor/sidekiq/worker_retries_exhausted_reporter.rb]: lol I love how I put things in lib you suggest they go to labor, I put things in labor you suggest they go to lib. Aye, one day I will get it right

File: app/labor/sidekiq/worker_retries_exhausted_reporter.rb (hunk 4 above)
Comment (mstruve) [reply app/labor/sidekiq/worker_retries_exhausted_reporter.rb]: I feel like this middleware's name is pretty self-explanatory. I am all for commenting on code when it is obscure but I feel like the class name and method are pretty straightforward.

**Comment by mstruve** [comment]
> what is "this number", exactly? we'll be alerted via a monitor whenever we have a worker that has failed 1 time? just want to make sure i understand the context here.

Any time we exhaust all retries for a job we will be alerted which means any time the "retries_exhausted" number goes above 1

File: config/initializers/sidekiq.rb (hunk 1 above)
Comment (mstruve) [reply config/initializers/sidekiq.rb]: `_ex` here is the exception. I added a comment with that link for more clarity

**Pull request merged**